FWCore/Concurrency/bin/edmStreamStallGrapher.py

```diff
@@ -46,94 +44,118 @@ def getTime(line):
+      processingSteps.append((name,trans,stream,time))
+      if stream > numStreams:
+        numStreams = stream
+    if l.find("event delayed read from source") != -1:
+      time = getTime(l)
+      if startTime == 0:
+        startTime = time
+      time = time - startTime
+      trans = kStarted
+      stream = 0
+      delayed = False
+      if l.find("finished:") != -1:
+        trans = kFinished
+      streamIndex = l.find("stream = ")
+      stream = int( l[streamIndex+9:l.find(" ",streamIndex+10)])
+      name = kSourceDelayedRead
+      if len(name) > maxNameSize:
+        maxNameSize = len(name)
+      processingSteps.append((name,trans,stream,time))
+      if stream > numStreams:
+        numStreams = stream
+
   f.close()
   return (processingSteps,numStreams,maxNameSize)
 
 
 #----------------------------------------------
 # Patterns:
 #
-# source: The source is identified by having a 'starting: delayed' followed immediately by a 'finished: delayed'
-#   the total time is actually the wait time plus the time the source was doing work
-# scheduled module: This is identified by no delays before call to 'starting' and then immediate call to 'finished'
-# unscheduled module: Has a 'starting: delayed' followed by a 'starting' followed by a 'finished' the a 'finished: delayed'
+# source: The source just records how long it was spent doing work, not how long it was stalled. Can get a lower
+#   bound on the stall time by measuring the time the stream was doing no work up till the source was run.
+# modules: time between prefetch finished and 'start processing' is the time it took to acquire any resources
 #
 #
 def findStalledModules(processingSteps, numStreams):
   streamTime = [0]*(numStreams+1)
   stalledModules = {}
   previousStartWasADelayed = [False]*(numStreams+1)
-  for n,trans,s,time,delayed in processingSteps:
+  modulesActiveOnStream = [ {} for x in xrange(0,numStreams+1)]
+  for n,trans,s,time in processingSteps:
     waitTime = None
+    modulesOnStream =modulesActiveOnStream[s]
+    if trans == kPrefetchEnd:
+      modulesOnStream[n]=time
     if trans == kStarted:
-      if delayed or (n == kSourceFindEvent):
-        previousStartWasADelayed[s] = True
-        streamTime[s] = time
-      else:
-        previousStartWasADelayed[s] = False
-        waitTime = time - streamTime[s]
-    else:
-      if (delayed or (n == kSourceFindEvent)) and previousStartWasADelayed[s]:
-        #for a source we only know the combined time for waiting and running
-        if n != kSourceFindEvent:
-          n = "source"
-        waitTime = time - streamTime[s]
-        previousStartWasADelayed[s] = False
+      if n in modulesOnStream:
+        waitTime = time - modulesOnStream[n]
+      if n == kSourceDelayedRead:
+        if 0 == len(modulesOnStream):
+          waitTime = time - streamTime[s]
+    if trans == kFinished:
+      if n != kSourceDelayedRead and n!=kSourceFindEvent and n!=kFinishInit:
+        del modulesOnStream[n]
```

Comment:
should I be able to run this file on a tracer output file generated in CMSSW_8_1_X_2016-08-18-1100?
I'm getting an error
```
edmStreamStallGrapher.py", line 385, in <module>
    stalledModules = findStalledModules(processingSteps, numStreams)
edmStreamStallGrapher.py", line 158, in findStalledModules
    del modulesOnStream[n]
KeyError: 'siPixelDigis'
```